pkg/oc/admin/prune/imageprune/worker.go

```diff
@@ -0,0 +1,349 @@
+		case imagegraph.ImageComponentTypeManifest:
+			return fmt.Sprintf("failed to delete image manifest link %s%s: %v", t.Component, detail, pf.Err)
+		default:
+			return fmt.Sprintf("failed to delete %s%s: %v", t.String(), detail, pf.Err)
+		}
+	default:
+		return fmt.Sprintf("failed to delete %v: %v", t, pf.Err)
+	}
+}
+
+// JobResult is a result of job's processing.
+type JobResult struct {
+	Job       *Job
+	Deletions []Deletion
+	Failures  []Failure
+}
+
+func (jr *JobResult) update(deletions []Deletion, failures []Failure) *JobResult {
+	jr.Deletions = append(jr.Deletions, deletions...)
+	jr.Failures = append(jr.Failures, failures...)
+	return jr
+}
+
+// Worker knows how to prune image and its related components.
+type Worker interface {
+	// Run is supposed to be run as a go-rutine. It terminates when nil is received through the in channel.
+	Run(in <-chan *Job, out chan<- JobResult)
+}
+
+type worker struct {
+	algorithm       pruneAlgorithm
+	registryClient  *http.Client
+	registryURL     *url.URL
+	imagePruner     ImageDeleter
+	streamPruner    ImageStreamDeleter
+	layerLinkPruner LayerLinkDeleter
+	blobPruner      BlobDeleter
+	manifestPruner  ManifestDeleter
+}
+
+var _ Worker = &worker{}
+
+// NewWorker creates a new pruning worker.
+// TODO: accept deleter factories rather than deleters themselves
+func NewWorker(
+	algorithm pruneAlgorithm,
+	registryClientFactory RegistryClientFactoryFunc,
+	registryURL *url.URL,
+	imagePruner ImageDeleter,
+	streamPruner ImageStreamDeleter,
+	layerLinkPruner LayerLinkDeleter,
+	blobPruner BlobDeleter,
+	manifestPruner ManifestDeleter,
+) (Worker, error) {
+	client, err := registryClientFactory()
+	if err != nil {
+		return nil, err
+	}
+
+	return &worker{
+		algorithm:       algorithm,
+		registryClient:  client,
+		registryURL:     registryURL,
+		imagePruner:     imagePruner,
+		streamPruner:    streamPruner,
+		layerLinkPruner: layerLinkPruner,
+		blobPruner:      blobPruner,
+		manifestPruner:  manifestPruner,
+	}, nil
+}
+
+func (w *worker) Run(in <-chan *Job, out chan<- JobResult) {
+	for {
+		job := <-in
+		if job == nil {
+			return
+		}
+		out <- *w.prune(job)
+	}
+}
+
+func (w *worker) prune(job *Job) *JobResult {
+	res := &JobResult{Job: job}
+
+	// If namespace is specified prune only ImageStreams and nothing more.
+	if len(w.algorithm.namespace) > 0 {
+		return res
+	}
+
+	if w.algorithm.pruneRegistry {
+		res.update(pruneImageComponents(
+			w.registryClient,
+			w.registryURL,
+			job.Components,
+			w.layerLinkPruner,
+		))
+
+		res.update(pruneBlobs(
+			w.registryClient,
+			w.registryURL,
+			job.Components,
+			w.blobPruner,
+		))
+
+		res.update(pruneManifests(
+			w.registryClient,
+			w.registryURL,
+			job.Components,
+			w.manifestPruner,
+		))
+	}
+
+	if len(res.Failures) > 0 {
+		// TODO: include image as a failure as well for the sake of summary's completness
+		return res
```

Comment:
True, I'm keeping the current behaviour that preserves image objects for future prunes so that blobs that failed to be pruned are recognized on the subsequent prunes. If we remove the image, we loose the option to prune its blobs using this kind of pruner.
But since we have a hard-pruner now, we could do the removal. I just hesitate to make it a default behaviour. Maybe a flag `--keep-broken-images` could be useful here. WDYT?

Reply:
A scenario speaking against the image deletion regardless of prio failure is that the registry lost write permission to the storage and deletion of each blob fails (already happened to some customers). In this case all images are deleted but all the blobs remain on the storage.
Is it worse than broken images? Hard to tell. It depends on what the customer wants to achieve - either he's running out of storage space or etcd is too big and slow or even both.
Therefor I'm more inclined to having a flag like `--keep-broken-images-on-failure` defaulting to `true` (current behaviour) so that the customer can choose what he prefers.

@bparees thoughts?

Reply:
I'm just afraid that hard-prune is not a very popular solution to this problem and am reluctant to make it a mandatory post-routine.

Reply:
it won't be mandatory unless there are other issues in their registry.  if we can't delete layers due to storage issues, they're going to have to fix their storage issues anyway... running hard prune again after doing so seems like a small additional burden.

Is there some other case where you think it would be better to leave the image data in place if we cannot remove the blob data?

Reply:
> In this case all images are deleted but all the blobs remain on the storage.
> Is it worse than broken images?

@miminar No, broken images mush worse. We have few ways how to remove blobs completely or even restore images in the etcd, but we do not have tools to restore blobs in the storage to fix images. We have only diagnostic tool  which can help to find such broken images. User should remove image from etcd and re-push such image by hands to fix them. Until he does this, these  images will be broken. There will be errors in push and pull. It seems to me that this is a much worse scenario, because the cluster goes into an inconsistent state and the correction requires manual intervention.

Reply:
> What about keeping the image only if there were no successful blob deletions?

Went for this one. Please let me know if you have any objections.